### ISSUE TITLE
Account for Finance Company payments toward order Financed amount when calculating balance

### ISSUE THREAD

**Issue by kbower** [kbower/tickettest1#819]
_Imported from trac ticket 819.
Created by **kbower**
Opened in trac: Mar 19, 2011, 2:34:37 PM
Last modified in trac: Nov 12, 2012, 12:23:29 PM_
***
_by kbower:_
**~~Instead of the approach we were initially thinking about, we need to instead change the order total calculation to exclude financed amount as a credit for finalized / voided orders~~** (Also not sufficient)

Subject:         RE: order balance, ar_trn payments from financers
Date:         Mon, 14 Mar 2011 17:21:31 -0400
From:         Jesus Capo <Jesus@ElDoradoFurniture.com>
To:         Kent Bower <kb@retailarchitects.com>, Mariam Farach <Mariam@ElDoradoFurniture.com>

There is no singular way of determining the difference between a finance company payment and a regular customer payment.  Here at El Dorado, we created a special type of "Payment Type" in the "MOP" Table so that we know what is a Finance Company payment (or other types of payments as well).  In our case, it is a "CKF" Payment type.

Here is a sampling of our payment types:

CKF Finance Company Check
CCK Cashier's Check
COR Corporate Check
TCK Travelers Check
BKT Bank/Wire Transfer
PRC Employee Payroll Deduction
STF State Farm Insurance Check
CKE Prsnl Ck, Elctrnc Deposit
ACH ACH CREDIT - BANK
VSI VISA-Internet
MCI MasterCard-Internet
DBV VISA Check Card
DBM MasterCard Check Card
MON Money Order
DBD Discover Check Card
AMZ AMERICAN EXPRESS - SLM
DIZ DISCOVER - SLM
DNZ DINERS CLUB - SLM
MCZ MASTERCARD - SLM
VSZ VISA - SLM
AMI American Express - Internet Payment
DII Discover Card - Internet Payment
CCU Credit Union Check
CCC Credit Card Convenience Check
FBC Foreign Bank Check
CKR Refund - CK Same Check Returned to Cust
AMP American Express - Phone Pyment
VSP VISA - Phone Payment
MCP MasterCard - Phone Pmt
DIP Discover Card - Phone Payment
CKO Check over $25,000 - Deposit
PPL Internet Paypal Account

-----Original Message-----
From: Kent Bower [mailto:kb@retailarchitects.com]
Sent: Monday, March 14, 2011 3:05 PM
To: Mariam Farach; Jesus Capo
Subject: order balance, ar_trn payments from financers

Mariam or JR,
        I'm peeling an onion which many layers to restructure being able to make a single payment and apply it to several orders.

I've got a question concerning an order's "Balance":

We sum the ar trn records to calculate the balance. However, I'd like to account for what has been financed as well.  If the financer has made one or more payments, I need to ignore those if I am using the SO.ORIG_FI_AMT.

Which AR_TRN records should I ignore (how do I determine the ar tran records that are payments from the finance company)?

I'm not sure this is making sense, so here is an example:

Original order total 'SAL' AR_TRN: $550
Paid cash: $50
SO.ORIG_FI_AMT: $500

Customer comes back and adds a second chair:

Total updated 'SAL': $1050
customer PMT: $50
SO.ORIG_FI_AMT: $500

Now, suppose the finance company pays $300 to the retailer:

'SAL' AR_TRN: $1050
customer PMT AR_TRN: $50
SO.ORIG_FI_AMT: $500
financer PMT AR_TRN: $300

When I calculate the balance for this order, I'd like to take the SAL total $1050, subtract the amount financed ($500) and all customer PMTs ($50) so I could display this order as having a remaining $500 balance.

How do identify the financer's $300 AR_TRN and being connected with the SO.ORIG_FI_AMT, so I can ignore it?  If I don't ignore it, I will get the wrong answer, showing a balance of only $200 instead of $500.

Kent

**Comment by kbower** [comment]
_**kbower** trac comment on Mar 22, 2011, 8:09:07 PM:_

Kent,

You are correct in that other retailers could be handling this differently. If they are using the system with the financing module then what happens is that you have a transfer of liability where upon finalization there is a dollar transfer between the customer’s account and the finance company therefore showing a zero balance on the customer’s account and a balance on the finance company account then the payment from the finance company is posted to their account.

Thanks,

Wayne Liburd
----
From: Kent Bower [mailto:kb@retailarchitects.com]
Sent: Monday, March 14, 2011 11:33 AM
To: Wayne Liburd
Subject: Re: Order Balance

I do not suppose we (retailarchitects) can assume financing means there is no balance due. 

Although Kane's submits the finance amount only after finalization, I suppose that is not necessarily what all retailers do. 

Even if all retailers operate the same, finance companies that issue an approval may not be fine with that amount changing after the approval.  Apparently GE MONEY is, but what if the customer later adds a second chair or entire living room set to the order?  What if they do not finance the full order amount?  (Even if Kane's generally doesn't do that, another retailer may.)

So our system cannot assume financing means there is no balance due.

For other retailers, will the AR_TRN.MOP_CD always be the same as the company code?  Is that a GERS implementation or is that the way Kane's has it set up? 

Is there anyway to tell that an AR_TRN payment is related to a finance?

Thanks Wayne,
Kent
----
Kent,

Why would you need to ignore one or more payments if you are using SO.ORIG_FI_AMT?

We generally assume if the order is financed there is no balance due. So if the customer makes any payments after the original finance amount it does not matter because when the sale is finalized we will submit only the difference between the sale amount and any deposits or payments made to the finance company.

The finance company payments for GE and made with AR_TRN.MOP_CD = 'GE' and 'CSH_DWR_CD = 'GE' however that will not happen until the sale is finalized and submitted for payment.

Thanks,

Wayne Liburd
----
Wayne, when determining an order's "Balance" we sum the ar trn records.

However, I'd like to account for what has been financed as well.

However, if the financer has made one or more payments, I need to ignore those if I am using the SO.ORIG_FI_AMT.
Which AR_TRN records should I ignore (how do I determine the ar tran records that are payments from the finance company)?

Kent

**Comment by kbower** [comment]
_**kbower** trac comment on Oct 13, 2011, 1:32:43 PM:_

It is quite possible that these finance company payments are _never_ applied to the actual order, but rather to the customer account (see accountnumber field of SAL `ArTran`).

In which case this ticket is closed.

**Comment by kbower** [comment]
_**kbower** trac comment on Oct 13, 2011, 3:09:59 PM:_

payments where tran.paymenttype.paymenttypetype = 'XB'
?

**Comment by kbower** [comment]
_**rneville** trac comment on Jul 26, 2012, 2:48:08 PM:_

this happens with voided orders too

**Comment by kbower** [comment]
_**kbower** trac comment on Jul 26, 2012, 3:35:14 PM:_

I need to be involved in the design review for this ticket, please.

**Comment by kbower** [comment]
_**kbower** trac comment on Oct 16, 2012, 1:42:37 PM:_

Ticket [#1965](1965) has been marked as a duplicate of this ticket.

**Comment by kbower** [comment]
_**kbower** trac comment on Oct 16, 2012, 4:45:45 PM:_

(Changeset [[3271]](../commit/9a40c9514ca4636637542463416473d9f75a8341))
 * Fixes [#1982](1982) (blank order balance in Customer History and inability to take payments against orders) by reverting [[3256]](../commit/2b961cb5e3530ab6ae8178296217be4050e3bde3) (see [#1965](1965))
   * relocate controller logic back in controller (runningbalance)
   * fix invalid assumption that archived transactions are always older than unarchived transactions by putting sort back in _join_transactions() again.
   * remove an unneeded reverse of transactions (efficiency, confusion)
   * unfortunately need to remove 'custhist_strict_filter' SystemParameter:
     * we can't be changing the orm's definition of a related collection by removing items it thinks should be there. 
       * the attribute becomes ambiguous, based on whether filtering took place or 'expire' was invoked, etc.
       * we've attempted mismatching orm definition with model definition in the past and failed on many accounts
     * filtering by mismatched customerid is too broad anyway, need to target 3rd party finance transactions more exactly (see #819)
     * real problem wasn't addressed anyway, which lies in the need for both a customer/salesperson balance and retailer balance (see #819)
 * remove two unreferenced flex Order properties

**Comment by kbower** [comment]
_**rneville** trac comment on Oct 31, 2012, 11:31:21 AM:_

Several steps for solving this:
 * create a new component for the order edit screen to replace the 'Balance' label and balance amt currently shown. The label will be changed to be 'Balance Due' to semantically match what we are attempting to show (which is the customer's amount due).
 * this new component (see above) will show a breakdown of totals, tax delivery, etc. and will also show the "accounting balance" in addition to how it arrived at the current due. It will also display the amount financed.
 * Order grids (i.e. order query) will be altered to add two new columns: Balance and Financed. The existing 'Balance' column will be renamed to 'Due' and will reflect the same as the 'Balance Due' label on the order edit screen.
 * A new field will be added to finance companies which will indicate a payment type that represents a payment from a 3rd party

The 'Balance Due' or 'Due' for an order will be calculated as follows:
Where 'accounting balance' is => sum(debits) - sum(credits) where orderid==orderid...
if accounting balance >= 0 and accounting balance <= financed amt, then balance due = $0
if accounting balance >= 0 and accounting balance > financed amt, then balance due = accounting balance - financed amt
if accounting balance < 0, then balance due = accounting balance

**Comment by kbower** [comment]
_**rneville** trac comment on Nov 12, 2012, 12:23:29 PM:_

(Changeset [[3378]](../commit/e8a001cd2218dde4d6293b44b77048794ad4f2f3))
* fixes #819 (Account for Finance Company payments toward order Financed amount when calculating balance)
  * Renames Balance label on order detail to 'Due'. This amount represents the amount due by the customer.
  * Creates a custom tooltip for all balance labels on the order detail screen to show a cleaner breakdown of how balances are arrived at.
  * Tooltip for 'Due' column shows true accounting balance of order, outstanding 3rd party payments due prior to settlement, etc.
  * Customer history grids altered to show total sale, amount financed, true balance & customer's balance due
  * new field added to Finance Company screens to assign a payment type (MOP) that represents payments from that company towards order balances
 * references [#1620](1620) (improved transaction visibility for 3rd party payments of financed orders with transfer of liability). A lot of #819 was necessary as a foundation for this ticket, which is not complete yet.
 * fixes [#2022](2022) (misc issues with contract print out)

**Comment by kbower** [comment]
_**rneville** trac status change on Nov 12, 2012, 12:23:29 PM:_ **closed**
